v8/usr/man/man1/view2d.1
.TH VIEW2D 1
.SH NAME
view2d \- movie of a function f(x, y, t)
.SH SYNOPSIS
.B view2d
[ option ...  ]
file
.LP
.B regrid
[ option ...  ]
file
.LP
.B vdata
[ option ...  ]
file
.SH DESCRIPTION
.I View2d
displays a sequence of functional surfaces on a 5620 or color display.
On the 5620, the surface is ruled with a square mesh
and projected isometrically;
on a color display, the surface is viewed from above,
with height indicated by color.
The options are:
.TP .6i
.B \-T5
output on the 5620;
.B \-T5
is the default when TERM (see
.IR environ (7))
is 5620.
.TP
.B \-Tc
output on color display;
.B \-Tc
is the default when TERM is not 5620.
.TP
.B \-Tp
produces input for
.IR plot (1)
to produce contour plots.
.TP
.BI \-c n
use 
.I n
colors
.RI ( n =32
by default)
or contours (6 by default).
.TP
.BI \-p s
run movie for
.I s
seconds
.RI ( s =5
by default).
.PP
There are two supported ways to generate input for
.IR view2d .
.I Vdata
takes ASCII input and allows scattered data;
.I view2d(3)
can be called from a program that
already has data on a grid.
.PP
.I Regrid
changes the grid sizes in
.I view2d
binary files,
with options:
.TP 1i
.B \-b
adds a bar at the bottom of the image indicating time.
.TP
.BI \-n NX\fR[,\fINY\fR]
output grid is
.I NX
by
.I NY.
.TP
.BI \-f NF
output will have
.I NF
frames.
.TP
.B \-r
the plane of closest fit to the first frame
is subtracted from all the frames.
.TP
.BI \-m fmin,fmax
clips the data to the range
.I fmin,fmax.
.PP
One option applies to both \fIview2d\fP and \fIregrid\fP:
.TP .8i
.BI \-t TS\fR[,\fITE\fR]
displays frames from
.I TS
(default first frame time)
to
.I TE
(default last frame time).
.PP
.I Vdata
takes input in the following format and converts to input for
.IR view2d .
The first line of each frame has
the number of data points in the frame (an integer)
and the time (a floating point value).
The rest of the lines in the frame consist of x y z triples.
The order of points is irrelevant.
Every (floating point or integer) number must be white space separated.
The options are:
.TP .85i
.BI \-n NX\fR[,\fINY\fR]
output grid is
.I NX
by
.I NY.
.TP
.B \-c
use piecewise constant rather than linear interpolation.
.TP
.B \-s
scatterplot the data on a black background.
.TP
.B \-i
the data is already on a square grid,
in the order x and y ascending with x varying fastest.
.SH "SEE ALSO"
view2d(3), view2d(5)
.br
Exotic options for photography, color maps, and halftone
are described in
.IR /usr/lib/view2d/howto .


v8/usr/src/cmd/view2d/view2d.c
#include	<stdio.h>
#include	<signal.h>
#include	<sgtty.h>
#include	"/usr/jerq/include/jioctl.h"
#include	"view2d.h"

#define		MAX		32767

struct sgttyb sttybuf, sttysave;
char *zflag = "";
int shade = 0;
int grid = 1;
typedef struct Point
{
	int x, y;
} Point;

short z[4096];
double ztime;
double fmid, fscale;

int zmax = -1000000000, zmin = 1000000000;
int nsleep, ntime;
int nframes = 0;
int period = 5;
int floor = -BIG;
char *progname;
short timewarp;
double ts, te;
Rd2d rd;

main(argc, argv)
	char **argv;
{
	char buf[512];
	int zc, scale, m, i;
	short ragnorok, dim;
	short *nz, *p, *q;
	int fd;
	int cv, tmin, tmax;
	short nx, ny, fru, frv;
	char *s;
	char **av = argv;
	int ac = argc;
	extern char *getenv();
	int magic[4];

	timewarp = 0;
	progname = argv[0];
	cv = ((s = getenv("TERM")) == 0) || strcmp(s, "5620");
	ioctl(0, TIOCGETP, &sttysave);
	strcpy(buf, "level");
	for(argc--, argv++; *argv && (**argv == '-'); argv++)
		switch(argv[0][1])
		{
		case 'z':
			zflag = "-z";
			break;
		case 's':
			shade++;
			break;
		case 'g':
			grid = 0;
			break;
		case 'T':
			if(argv[0][2] == 'c') cv = 1;
			else if(argv[0][2] == '5') cv = 0;
			else if(argv[0][2] == 'p') cv = 2;
			else if(argv[0][2] == 'h') cv = 3;
			else goto err;
			break;
		case 't':
			i = sscanf(&argv[0][2], "%hd, %hd", &ts, &te);
			if(i==0) error("bad TS,TE");
			timewarp = i;
			break;
		case 'p':
			sscanf(&argv[0][2], "%d", &period);
			break;
		case 'i':
			sscanf(&argv[0][2], "%d", &floor);
			break;
		case 'B':
		case 'D':
		case 'S':
		case 'b':
		case 'm':
		case 'w':
			/* these options only apply in color, so assume -Tc */
			cv = 1;
			break;
		case 'c':
			if(cv==0) cv = 1; /* maybe -Tp, but probably -Tc */
		case 'v':
			break;
		default:
		err:
			fprintf(stderr, "Usage: view2d [-Tdev] [-ptime] [-cn] [-ms]\n");
			exit(1);
		}
	if(*argv){
		if((fd = open(*argv, 0)) == -1)
			quit("cannot open input");
	}else{
		fd = 0;  /* if no file argument, try standard input */
	}
	if(cv==1)
	{
		if(access((s="/usr/lib/view2d/level"), 1) == -1)
		{
			fprintf(stderr, "go to swift for color\n");
			exit(1);
		}
		dup2(fd,0);     /* close(0); dup(fd); close(fd); */
		strcpy(*av, "level");
		execvp(s, av);
		perror(s);
		exit(1);
	} else if(cv==2) {
		if(access((s="/usr/lib/view2d/contour"), 1) == -1)
		{
			fprintf(stderr, "can't find %s\n",s);
			exit(1);
		}
		dup2(fd,0);     /* close(0); dup(fd); close(fd); */
		strcpy(*av, "cntour");
			/* WARNING: strlen("cntour") must <= strlen("view2d") */
		execvp(s, av);
		perror(s);
		exit(1);
	} else if(cv==3)
	{
		if(access((s="/usr/lib/view2d/halft"), 1) == -1)
		{
			fprintf(stderr, "can't find %s\n",s);
			exit(1);
		}
		dup2(fd,0);     /* close(0); dup(fd); close(fd); */
		strcpy(*av, "halft");
		execvp(s, av);
		perror(s);
		exit(1);
	}
	sttybuf = sttysave;
	sttybuf.sg_flags |= RAW;
	sttybuf.sg_flags &= ~ECHO;
	ioctl(0, TIOCSETP, &sttybuf);
	rd2dh(fd, &nx, &ny);
	if(nx*ny*sizeof(short) > sizeof z)
	{
		fprintf(stderr, "nx*ny=%d: must be less than 4096\n", nx*ny);
		exit(1);
	}
	sprintf(buf, "/usr/jerq/bin/32ld %s /usr/lib/view2d/view2d.m", zflag);
	if(system(buf)) exit(1);
	putchar('X'); sendn(nx);
	putchar('Y'); sendn(ny);
	if(grid) putchar('G');
	if(shade) putchar('S');
	putchar('P');
	sendn(period);
	putchar('n');
	sendn(rd.nfr);
	putchar('I');
	sendn(floor);
	m = 2*nx*ny;
	fmid = (rd.fmin+rd.fmax)/2;
	fscale = MAX/(rd.fmax - fmid);
	while(rd2di(&ztime, z))
	{
		char buf[56];

		putchar('F');
		sprintf(buf, "%g", ztime);
		sends(buf);
		for(q = nx*ny + (p = z); q != p; p++)
			sendn((short)((*p-fmid)*fscale));
	}

	putchar('Q');
	fflush(stdout);
	(void)getchar();
	quit(0);
}

quit(s)
	char *s;
{
	ioctl(0, JTERM, 0);
	ioctl(0, TIOCSETP, &sttysave);
	if(s)
		printf("view2d: %s\n", s);
	exit(0);
}

sendn(n)
{
	putchar(n>>8);
	putchar(n);
}

sends(s)
	char *s;
{
	while(*s) putchar(*s++);
	putchar(0);
}

abs(n)
{
	return(n<0? -n:n);
}
